Wrap RedisStorage in CachedStorage

masap opened the pull request:
<!-- Thank you for creating a pull request! In general, we merge your pull requests after they get two or more approvals. -->

## Motivation
This PR fixes issue #2938.

Before this PR, ```study.optimize()``` required 305 seconds when ```n_trials=1000```, after this PR, it requires 128 seconds. This PR reduced the execution time about 58%.

<!-- Describe your motivation why you will submit this PR. This is useful for reviewers to understand the context of PR. -->

## Description of the changes
<!-- Describe the changes in this PR. -->
Previously ```CachedStorage``` only wrap the ```RDBStorage```. This PR wrap RedisStorage in CachedStorage also.

Signed-off-by: Masashi Honma <masashi.honma@gmail.com>

masap commented:
This is the code used for measurement.

```python
import optuna
import time

def objective(trial):
    x = trial.suggest_float('x', -10, 10)
    return (x - 2) ** 2

res = []
for i in range(10):
    storage = optuna.storages.RedisStorage(
        url='redis://localhost:6379',
    )

    study = optuna.create_study(storage=storage)

    start_time = time.monotonic()
    study.optimize(objective, n_trials=1000)
    end_time = time.monotonic()

    res.append(end_time - start_time)

print(res)
```

HideakiImamura commented:
@keisuke-umezawa Could you review this PR since you are the author of #2938?

HideakiImamura commented on optuna/storages/_redis.py:
```diff
@@ -321,6 +322,24 @@ def get_all_study_summaries(self) -> List[StudySummary]:
+    def _create_new_trial(
+        self, study_id: int, template_trial: Optional[FrozenTrial] = None
+    ) -> FrozenTrial:
+        """Create a new trial and returns its trial_id and a :class:`~optuna.trial.FrozenTrial`.
```
For consistency.
```suggestion
        """Create a new trial and returns a :class:`~optuna.trial.FrozenTrial`.
```

masap replied:
Thanks! I also fixed same typo in ```RDBStorage```.

HideakiImamura commented on optuna/storages/_redis.py:
```diff
@@ -104,7 +105,7 @@ def __init__(
         self._heartbeat_interval = heartbeat_interval
         self._grace_period = grace_period
-        self._failed_trial_callback = failed_trial_callback
+        self.failed_trial_callback = failed_trial_callback
```
Why do you change the name of this attribute?

masap replied:
Because previously we used ```_failed_trial_callback``` only for ```RedisStorage```, but now it is also used by ```_CachedStorage```.

masap replied:
Now I modifyed ```get_failed_trial_callback``` of ```_CachedStorage``` to call ```get_failed_trial_callback``` of ```_backend``` instead. Because it is more reasonable and increases test coverage.

HideakiImamura replied:
Sounds reasonable.

HideakiImamura commented on optuna/storages/_redis.py:
```diff
@@ -505,6 +524,18 @@ def _set_best_trial(self, study_id: int, trial_id: int) -> None:
             pipe.set(self._key_study_summary(study_id), pickle.dumps(study_summary))
             pipe.execute()
 
+    def _check_and_set_param_distribution(
```
This method is used in `_CachedStorage`. I think we should define this method in the `BaseStorage` for the maintainability. What do you think?

masap replied:
I do not think so...
Because ```BaseStorage``` is also inhedited by ```InMemoryStorage``` and it does not use ```_check_and_set_param_distribution```.
```
BaseStorage
 |
 | inherit
 |
 +-- _CachedStorage (calls _check_and_set_param_distribution of wrapping class)
 |    |
 |    | wrap
 |    |
 |    +--- RDBStorage
 |    |
 |    +--- RedisStorage
 |
 +-- InMemoryStorage (does not calls _check_and_set_param_distribution because InMemoryStorage does not wrapping class)
```

keisuke-umezawa replied:
Even though it will not be used by `InMemoryStorage`, we can define it in `BaseStorage`. Also, it will be better if we will define another wrapped storeges.
Could you move the definition to `BaseStorage`?

masap replied:
That fix seems to be beyond the scope of this PR... So I will fix it in another PR after this PR is merged. Is it OK ?

HideakiImamura reviewed (commented):
Thanks for the PR! I have several early comments. Could you take a look?

masap commented:
> Besides below comments, it looks like the test coverage decreases. Do you have any idea to improve it?

Ok, I will try to fix it.

masap commented:
When used with ```_CachedStorage```, these functions in ```RedisStorage``` are not used.
```
get_trial_number_from_id
read_trials_from_remote_storage
fail_stale_trials
```
These functions decrease test coverage. Of course we can add some unit tests for them to increase test coverage. However, since we will not use ```RedisStorage``` directly in the future, but will use it together with ```_CachedStorage```, there is no point in keeping these functions. So I changed these functions to raise ```NotImplementedError```. ```RDBStorage```, which has a similar problem, has also been fixed.

Any comments would be appreciated.

keisuke-umezawa commented:
@masap 
Thank you for the implementation!

> get_trial_number_from_id
> read_trials_from_remote_storage
> fail_stale_trials
> These functions decrease test coverage. Of course we can add some unit tests for them to increase test coverage. However, since we will not use RedisStorage directly in the future, but will use it together with _CachedStorage, there is no point in keeping these functions. So I changed these functions to raise NotImplementedError. RDBStorage, which has a similar problem, has also been fixed.

Basically, this is true. But, we have not concluded that we will not provide any chances for users to use RedisStorage and RDBStorage directly. Maybe, we will concluded it like the above. But, before that, could you leave those functions as they are?

masap commented:
@keisuke-umezawa Thank you for your review! Ok, I'll leave those functions as they were.

masap commented:
I reverted these functions, but that reduces the test coverage, so I added some tests.
But now the coverage is not good still, I will investigate it.

masap commented:
Now test coverage is fixed.
Any comments would be appreciated.

HideakiImamura commented on tests/storages_tests/test_storages.py:
```diff
@@ -158,7 +158,6 @@ def test_get_study_id_from_name_and_get_study_name_from_id(storage_mode: str) ->
         # Generate unique study_name from the current function name and storage_mode.
         function_name = test_get_study_id_from_name_and_get_study_name_from_id.__name__
         study_name = function_name + "/" + storage_mode
-        storage = optuna.storages.get_storage(storage)
```
Why this line is deleted? Many other calls to get_storage have also been removed, why?

masap replied:
英語で正確に説明する自信が無かったので日本語にて失礼します。

この PR は ```RedisStorage``` を ```CachedStorage``` で wrap しており、この修正は test coverage を悪化させます。理由は、いくつかの関数（今まで直接 ```RedisStorage``` を使用していた場合は呼ばれていた関数）が呼ばれなくなるからです。

```RDBStorage``` についても同様に、```CachedStorage``` で wrap した際に、同じ関数が呼ばれなくなります。

最初の commit においては、 test coverage が悪化することは認識していたものの、 ```RDBStorage``` と同じレベルに悪化するので問題ないだろうと思い、そのままの状態で PR を送りました。

しかし、 test coverage を改善するようにとのご指摘があり、改善のために、 ```RDBStorage/RedisStorage``` の呼ばれない関数を obsolete にしてみました。

しかし、 ```RDBStorage/RedisStorage``` を ```CachedStorage``` で wrap せずに使う可能性があるので、それはやめて欲しいとのご指摘がありました。

そこで、 test coverage の改善のために、 ```RDBStorage/RedisStorage``` を直接使用するテストケースを追加しました。いくつかのテストにおいて ```get_storage()``` を削除しているのはそのためです。```get_storage()``` は ```CachedStorage``` で ```RDBStorage/RedisStorage``` を wrap してしまうので、それを避ける必要がありました。

masap replied:
This PR wraps `RedisStorage` with `CachedStorage`, and this PR worsens the test coverage. The reason is that some functions (that would have been called if you were using `RedisStorage` directly) will no longer be called.

Similarly for `RDBStorage`, the same functions will no longer be called when wrapping with `CachedStorage`.

In the first commit, I was aware that the test coverage would deteriorate, but I thought it would be fine since it would deteriorate to the same level as `RDBStorage`, so I sent the PR as is.

However, I was asked to improve the test coverage, so I tried to obsolete the functions that are not called in `RDBStorage/RedisStorage` in order to improve it.

However, it was pointed out that `RDBStorage/RedisStorage` may be used without wrapping in `CachedStorage`, and they don't want that to happen.

So, to improve the test coverage, I have added test cases that use `RDBStorage/RedisStorage` directly. That's why I removed `get_storage()` in some tests, because `get_storage()` wraps `RDBStorage/RedisStorage` in `CachedStorage`, and we needed to avoid that.

HideakiImamura replied:
Thank you for the explanation. I agree with you.

HideakiImamura commented on tests/storages_tests/test_storages.py:
```diff
@@ -1232,3 +1227,28 @@ def failed_trial_callback(study: "optuna.Study", trial: FrozenTrial) -> None:
+
+
+@pytest.mark.parametrize("storage_mode", ["sqlite", "redis"])
+def test_fail_stale_trials_raw(storage_mode: str) -> None:
```
What is the purpose of this test?

masap replied:
```study``` を作成する際に内部で必ず ```get_storage()``` を呼んでしまうので ```fail_stale_trials()``` を生の ```RDBStorage/RedisStorage``` でテストすることは現状できません。

そこで、苦肉の策として生の ```RDBStorage/RedisStorage``` から ```fail_stale_trials()``` を呼んで、エラーにならないことを確認するレベルのテストがこれです。

「これだったら無い方がマシ」ということであれば削除します。

masap replied:
Currently, it is not possible to test `fail_stale_trials()` on raw `RDBStorage/RedisStorage` because `get_storage()` is always called internally when the `study` is created.

So, as a "last resort", here is a test that calls `fail_stale_trials()` from raw `RDBStorage/RedisStorage` to make sure it doesn't fail. If you think you're better off without it, I will remove it.

HideakiImamura replied:
Thank you for the explanation. How about adding the comment to note that we should remove this test if we have a option of the `get_storage` to enable a cache?

masap replied:
Ok, fixed.

HideakiImamura commented on tests/storages_tests/test_storages.py:
```diff
@@ -1232,3 +1227,28 @@ def failed_trial_callback(study: "optuna.Study", trial: FrozenTrial) -> None:
+        assert storage.fail_stale_trials(study_id) == []
+
+
+@pytest.mark.parametrize("storage_mode", STORAGE_MODES_HEARTBEAT)
```
Why the possible `storage_mode`s are `STORAGE_MODES_HEARTBEAT`? Might be `STORAGE_MODES`.

masap replied:
```STORAGE_MODES``` は ```inmemory``` を含みますが、```inmemory```はもちろんremote storageではないので```read_trials_from_remote_storage()```を持たず、エラーとなります。

そのため```STORAGE_MODES_HEARTBEAT```を使用しました。「hertbeat と無関係なテストでなぜ```STORAGE_MODES_HEARTBEAT```を使うのだ？」という違和感があるのは理解できます。私もそうだったので、当初は ```"sqlite", "cached_sqlite", "redis", "cached_redis"``` と並べて書こうかなとも思いましたが、結局「定数使え」と言われそうだなと思い、今の形にしました。

masap replied:
`STORAGE_MODES` contains `inmemory` storage, but `inmemory` storage is of course not remote storage, so it does not have `read_trials_from_remote_storage()`, which will result in an error.

This is why we used `STORAGE_MODES_HEARTBEAT`. You may be wondering, "Why use `STORAGE_MODES_HEARTBEAT` for a test that has nothing to do with hertbeat?" I can understand why you might feel uncomfortable. At first, I thought about writing `"sqlite", "cached_sqlite", "redis", and "cached_redis"`, but then I thought that people would say "use constants", so I decided to use the current form.

HideakiImamura replied:
Thank yoo for the explanation. I agree with the idea that we should not test this with the in-memory storage. How about using `STORAGE_MODE` and early return the test if the `storage_mode` is the in-memory storage? It would be appriciate that you provide some useful comments.

masap replied:
Sorry. This is my mis-undestanding. `InMemoryStorage` has `read_trials_from_remote_storage()` and it can pass the test. So I fixed it to use just `STORAGE_MODES`.

I have modified `read_trials_from_remote_storage()` of `InMemoryStorage` to raise `NotImplemented` exception for investigation and wrongly used it for test...

HideakiImamura reviewed (commented):
Thanks for the update. I have several comments around tests. Please take a look.

HideakiImamura reviewed (approved):
Thanks for the fix! LGTM!

himkt commented on optuna/integration/pytorch_lightning.py:
```diff
@@ -60,7 +61,9 @@ def on_init_start(self, trainer: Trainer) -> None:
             if version.parse(pl.__version__) < version.parse("1.5.0"):
                 raise ValueError("PyTorch Lightning>=1.5.0 is required in DDP.")
-            if not isinstance(self._trial.study._storage, _CachedStorage):
+            if (not isinstance(self._trial.study._storage, _CachedStorage)) or (
+                not isinstance(self._trial.study._storage._backend, RDBStorage)
+            ):
```
```suggestion
            if not (
                isinstance(self._trial.study._storage, _CachedStorage)
                and isinstance(self._trial.study._storage._backend, RDBStorage)
            ):
```

How about using transformed version?

masap replied:
Thanks, it is more clear. Fixed.

keisuke-umezawa commented on optuna/storages/_redis.py:
```diff
@@ -618,20 +649,31 @@ def get_all_trials(
+        else:
+            return trials
+
+    def _get_trials(
```
Could you add unit tests for this method? esp, for exlucded_trial_ids.

masap replied:
Ok, fixed. I reused the `test_get_trials_excluded_trial_ids()` of `RDBStorage` test.

keisuke-umezawa reviewed (commented):
Almost LGTM. Could you check a few comments?

masap commented:
I also fixed some existing `black` and `blackdoc` errors to avoid check failure.

masap commented:
I fixed `black` and `blackdoc` errors at #3260. Please re run test for this PR after #3260 is merged.

masap commented:
Rebased.

keisuke-umezawa reviewed (approved):
LGTM! Thank you for your contribution!
